app/services/analytics_service.rb

```diff
@@ -0,0 +1,102 @@
+    else
+      @reaction_data ||= Reaction.where(reactable_id: article_ids, reactable_type: "Article").where("points > 0")
+      @comment_data ||= Comment.where(commentable_id: article_ids, commentable_type: "Article").where("score > 0")
+      @follow_data ||= Follow.where(followable_type: user_or_org.class.name, followable_id: user_or_org.id)
+      @page_view_data ||= PageView.where(article_id: article_ids)
+    end
+  end
+
+  def totals
+    total_views = article_data.sum(:page_views_count)
+    logged_in_page_view_data = page_view_data.select { |pv| pv.user_id.present? }
+    average_read_time_in_seconds = logged_in_page_view_data.size.positive? ? logged_in_page_view_data.sum(&:time_tracked_in_seconds) / logged_in_page_view_data.size : 0
+
+    {
+      comments: {
+        total: comment_data.size
+      },
+      reactions: {
+        total: reaction_data.size,
+        like: reaction_data.count { |rxn| rxn.category == "like" },
+        readinglist: reaction_data.count { |rxn| rxn.category == "readinglist" },
+        unicorn: reaction_data.count { |rxn| rxn.category == "unicorn" }
+      },
+      follows: {
+        total: follow_data.size
+      },
+      page_views: {
+        total: total_views,
+        average_read_time_in_seconds: average_read_time_in_seconds,
+        total_read_time_in_seconds: average_read_time_in_seconds * total_views
+      }
+    }
+  end
+
+  def stats_grouped_by_day
+    final_hash = {}
+
+    dates.each do |date|
+      string_date = date.strftime("%a, %m/%d")
+      reaction_data_of_date = reaction_data.select { |rxn| rxn.created_at.beginning_of_day.to_i == date.to_i }
+      logged_in_page_view_data = page_view_data.select { |pv| pv.created_at.beginning_of_day.to_i == date.to_i && pv.user_id.present? }
+      average_read_time_in_seconds = logged_in_page_view_data.size.positive? ? logged_in_page_view_data.sum(&:time_tracked_in_seconds) / logged_in_page_view_data.size : 0
+      total_views = page_view_data.select { |pv| pv.created_at.beginning_of_day.to_i == date.to_i }.sum(&:counts_for_number_of_views)
+
+      final_hash[string_date] = {
+        comments: {
+          total: comment_data.count { |c| c.created_at.beginning_of_day.to_i == date.to_i }
+        },
+        reactions: {
+          total: reaction_data_of_date.count,
+          like: reaction_data_of_date.count { |rxn| rxn.category == "like" },
+          readinglist: reaction_data_of_date.count { |rxn| rxn.category == "readinglist" },
+          unicorn: reaction_data_of_date.count { |rxn| rxn.category == "unicorn" }
+        },
+        page_views: {
+          total: total_views,
+          total_read_time_in_seconds: average_read_time_in_seconds * total_views,
+          average_read_time_in_seconds: average_read_time_in_seconds
+        },
+        follows: {
+          total: follow_data.count { |f| f.created_at.beginning_of_day.to_i == date.to_i }
+        }
+      }
+    end
+
+    final_hash
+  end
+
+  def start_date
```

Comment:
Does every method need to be public? I also suggest making `attr_readers` private (those that don't need to be public).

Reply:
Made date methods and `attr_readers` private 👍